File: benchmarks-website/server/src/slug.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

//! Opaque slugs for `/api/chart/:slug`.
//!
//! Per `02-contracts.md`, the web-ui treats slugs as opaque strings: it
//! receives them from `/api/groups` and feeds them back unchanged to
//! `/api/chart/:slug`. The server is free to choose any format.
//!
//! Slugs here are `<prefix>.<base64url-of-json>` where `<prefix>` names the
//! source fact table and the JSON encodes the chart key. Round-tripping the
//! slug back gives a strongly-typed [`ChartKey`].

use anyhow::Context as _;
use anyhow::Result;
use anyhow::anyhow;
use base64::Engine as _;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use serde::Deserialize;
use serde::Serialize;

const PREFIX_QUERY: &str = "qm";
const PREFIX_COMPRESSION_TIME: &str = "ct";
const PREFIX_COMPRESSION_SIZE: &str = "cs";
const PREFIX_RANDOM_ACCESS: &str = "rat";
const PREFIX_VECTOR_SEARCH: &str = "vsr";

const PREFIX_QUERY_GROUP: &str = "qmg";
const PREFIX_COMPRESSION_TIME_GROUP: &str = "ctg";
const PREFIX_COMPRESSION_SIZE_GROUP: &str = "csg";
const PREFIX_RANDOM_ACCESS_GROUP: &str = "rag";
const PREFIX_VECTOR_SEARCH_GROUP: &str = "vsg";

/// The strongly-typed chart key parsed from a slug.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "k")]
pub enum ChartKey {
    /// `query_measurements` chart: `(dataset, query_idx)` per `01-schema.md`.
    /// Group context (`dataset_variant`, `scale_factor`, `storage`) is carried
    /// alongside so the slug fully specifies the chart.
    QueryMeasurement {
        dataset: String,
        dataset_variant: Option<String>,
        scale_factor: Option<String>,
        storage: String,
        query_idx: i32,
    },
    /// `compression_times` chart: `(dataset, dataset_variant)`.
    CompressionTime {
        dataset: String,
        dataset_variant: Option<String>,
    },
    /// `compression_sizes` chart: `(dataset, dataset_variant)`.
    CompressionSize {
        dataset: String,
        dataset_variant: Option<String>,
    },
    /// `random_access_times` chart: `dataset`.
    RandomAccess { dataset: String },
    /// `vector_search_runs` chart: `(dataset, layout, threshold)`.
    VectorSearch {
        dataset: String,
        layout: String,
        threshold: f64,
    },
}

impl ChartKey {
    fn prefix(&self) -> &'static str {
        match self {
            Self::QueryMeasurement { .. } => PREFIX_QUERY,
            Self::CompressionTime { .. } => PREFIX_COMPRESSION_TIME,
            Self::CompressionSize { .. } => PREFIX_COMPRESSION_SIZE,
            Self::RandomAccess { .. } => PREFIX_RANDOM_ACCESS,
            Self::VectorSearch { .. } => PREFIX_VECTOR_SEARCH,
        }
    }

    /// Render the slug for this chart key.
    pub fn to_slug(&self) -> String {
        let json = serde_json::to_vec(self).expect("ChartKey is always JSON-serializable");
        format!("{}.{}", self.prefix(), URL_SAFE_NO_PAD.encode(json))
    }

    /// Parse a slug previously produced by [`Self::to_slug`].
    pub fn from_slug(slug: &str) -> Result<Self> {
        let (_, encoded) = slug
            .split_once('.')
            .ok_or_else(|| anyhow!("slug missing '.' separator"))?;
        let json = URL_SAFE_NO_PAD
            .decode(encoded.as_bytes())
            .context("slug payload was not valid base64url")?;
        let key: Self = serde_json::from_slice(&json).context("slug payload was not valid JSON")?;
        Ok(key)
    }
}

/// Slug for a *group* of charts. Mirrors [`ChartKey`] but at the group
/// granularity: a group is a set of charts that share every dimension except
/// one (e.g. all 22 TPC-H queries at sf=1 on nvme).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "k")]
pub enum GroupKey {
    /// All `query_measurements` charts at one `(dataset, dataset_variant,
    /// scale_factor, storage)`. Charts inside vary by `query_idx`.
    QueryGroup {
        dataset: String,
        dataset_variant: Option<String>,
        scale_factor: Option<String>,
        storage: String,
    },
    /// Every compression-time chart (one per `(dataset, dataset_variant)`).
    CompressionTimeGroup,
    /// Every compression-size chart.
    CompressionSizeGroup,
    /// Every random-access chart.
    RandomAccessGroup,
    /// All vector-search charts at one `(dataset, layout)`. Charts inside
    /// vary by `threshold`.
    VectorSearchGroup { dataset: String, layout: String },
}

impl GroupKey {
    fn prefix(&self) -> &'static str {
        match self {
            Self::QueryGroup { .. } => PREFIX_QUERY_GROUP,
            Self::CompressionTimeGroup => PREFIX_COMPRESSION_TIME_GROUP,
            Self::CompressionSizeGroup => PREFIX_COMPRESSION_SIZE_GROUP,
            Self::RandomAccessGroup => PREFIX_RANDOM_ACCESS_GROUP,
            Self::VectorSearchGroup { .. } => PREFIX_VECTOR_SEARCH_GROUP,
        }
    }

    /// Render the slug for this group key.
    pub fn to_slug(&self) -> String {
        let json = serde_json::to_vec(self).expect("GroupKey is always JSON-serializable");
        format!("{}.{}", self.prefix(), URL_SAFE_NO_PAD.encode(json))
    }

    /// Parse a slug previously produced by [`Self::to_slug`].
    pub fn from_slug(slug: &str) -> Result<Self> {
        let (_, encoded) = slug
            .split_once('.')
            .ok_or_else(|| anyhow!("slug missing '.' separator"))?;
        let json = URL_SAFE_NO_PAD
            .decode(encoded.as_bytes())
            .context("slug payload was not valid base64url")?;
        let key: Self = serde_json::from_slice(&json).context("slug payload was not valid JSON")?;
        Ok(key)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn roundtrip(key: ChartKey) {
        let slug = key.to_slug();
        let parsed = ChartKey::from_slug(&slug).expect("parses back");
        assert_eq!(parsed, key);
    }

    #[test]
    fn query_measurement_roundtrips() {
        roundtrip(ChartKey::QueryMeasurement {
            dataset: "tpch".into(),
            dataset_variant: None,
            scale_factor: Some("1".into()),
            storage: "nvme".into(),
            query_idx: 7,
        });
    }

    #[test]
    fn vector_search_roundtrips() {
        roundtrip(ChartKey::VectorSearch {
            dataset: "cohere-large-10m".into(),
            layout: "partitioned".into(),
            threshold: 0.75,
        });
    }

    #[test]
    fn random_access_roundtrips() {
        roundtrip(ChartKey::RandomAccess {
            dataset: "taxi".into(),
        });
    }

    #[test]
    fn malformed_slug_rejected() {
        assert!(ChartKey::from_slug("not-a-slug").is_err());
        assert!(ChartKey::from_slug("qm.****").is_err());
    }

    fn roundtrip_group(key: GroupKey) {
        let slug = key.to_slug();
        let parsed = GroupKey::from_slug(&slug).expect("parses back");
        assert_eq!(parsed, key);
    }

    #[test]
    fn group_keys_roundtrip() {
        roundtrip_group(GroupKey::QueryGroup {
            dataset: "tpch".into(),
            dataset_variant: None,
            scale_factor: Some("1".into()),
            storage: "nvme".into(),
        });
        roundtrip_group(GroupKey::CompressionTimeGroup);
        roundtrip_group(GroupKey::CompressionSizeGroup);
        roundtrip_group(GroupKey::RandomAccessGroup);
        roundtrip_group(GroupKey::VectorSearchGroup {
            dataset: "cohere".into(),
            layout: "partitioned".into(),
        });
    }

    #[test]
    fn group_slug_prefix_distinct_from_chart() {
        let chart = ChartKey::CompressionTime {
            dataset: "tpch".into(),
            dataset_variant: None,
        }
        .to_slug();
        let group = GroupKey::CompressionTimeGroup.to_slug();
        let chart_prefix = chart.split_once('.').unwrap().0;
        let group_prefix = group.split_once('.').unwrap().0;
        assert_ne!(chart_prefix, group_prefix);
    }

    #[test]
    fn malformed_group_slug_rejected() {
        assert!(GroupKey::from_slug("not-a-slug").is_err());
        assert!(GroupKey::from_slug("qmg.****").is_err());
    }
}

File: benchmarks-website/web/lib/slug.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

/**
 * Opaque slugs for `/api/chart/{slug}` and `/api/group/{slug}`, the TypeScript
 * port of `server/src/slug.rs`.
 *
 * The web-ui treats slugs as opaque strings: it receives them from
 * `/api/groups` and feeds them back unchanged, so this read service both
 * produces and consumes them. A slug is `<prefix>.<base64url-of-json>`, where
 * `<prefix>` names the source fact table (from [`./families`]) and the JSON
 * encodes the chart or group key. The JSON keys are emitted in the same order
 * the Rust structs declare them, with the `k` discriminant first (serde's
 * internally-tagged representation), and `Option` fields are emitted as
 * explicit `null` rather than omitted, mirroring serde's default. Round-tripping
 * a slug back gives a strongly-typed [`ChartKey`] or [`GroupKey`].
 */

import { familyForChartKind, familyForGroupKind, type ChartKind, type GroupKind } from './families';

/** Strongly-typed chart key parsed from a slug. Discriminated on `k`. */
export type ChartKey =
  | {
      k: 'QueryMeasurement';
      dataset: string;
      dataset_variant: string | null;
      scale_factor: string | null;
      storage: string;
      query_idx: number;
    }
  | { k: 'CompressionTime'; dataset: string; dataset_variant: string | null }
  | { k: 'CompressionSize'; dataset: string; dataset_variant: string | null }
  | { k: 'RandomAccess'; dataset: string }
  | { k: 'VectorSearch'; dataset: string; layout: string; threshold: number };

/** Strongly-typed group key parsed from a slug. Discriminated on `k`. */
export type GroupKey =
  | {
      k: 'QueryGroup';
      dataset: string;
      dataset_variant: string | null;
      scale_factor: string | null;
      storage: string;
    }
  | { k: 'CompressionTimeGroup' }
  | { k: 'CompressionSizeGroup' }
  | { k: 'RandomAccessGroup' }
  | { k: 'VectorSearchGroup'; dataset: string; layout: string };

const CHART_KINDS: readonly ChartKind[] = [
  'QueryMeasurement',
  'CompressionTime',
  'CompressionSize',
  'RandomAccess',
  'VectorSearch',
];

const GROUP_KINDS: readonly GroupKind[] = [
  'QueryGroup',
  'CompressionTimeGroup',
  'CompressionSizeGroup',
  'RandomAccessGroup',
  'VectorSearchGroup',
];

function encodePayload(json: string): string {
  return Buffer.from(json, 'utf8').toString('base64url');
}

function decodePayload(slug: string): unknown {
  const dot = slug.indexOf('.');
  if (dot === -1) {
    throw new Error("slug missing '.' separator");
  }
  // The prefix before the '.' is ignored on decode (it is a redundant family
  // hint); the JSON payload is the source of truth, matching `slug.rs`.
  const encoded = slug.slice(dot + 1);
  const decoded = Buffer.from(encoded, 'base64url').toString('utf8');
  // An empty decode (e.g. all-invalid base64url) yields '', and `JSON.parse('')`
  // throws, so a malformed payload is rejected here.
  return JSON.parse(decoded) as unknown;
}

/**
 * Build the canonical JSON for a chart key with keys in the Rust declaration
 * order (`k` first), then render `<prefix>.<base64url-of-json>`.
 */
export function chartKeyToSlug(key: ChartKey): string {
  const prefix = familyForChartKind(key.k).chartSlugPrefix;
  let ordered: Record<string, unknown>;
  switch (key.k) {
    case 'QueryMeasurement':
      ordered = {
        k: key.k,
        dataset: key.dataset,
        dataset_variant: key.dataset_variant,
        scale_factor: key.scale_factor,
        storage: key.storage,
        query_idx: key.query_idx,
      };
      break;
    case 'CompressionTime':
    case 'CompressionSize':
      ordered = { k: key.k, dataset: key.dataset, dataset_variant: key.dataset_variant };
      break;
    case 'RandomAccess':
      ordered = { k: key.k, dataset: key.dataset };
      break;
    case 'VectorSearch':
      ordered = { k: key.k, dataset: key.dataset, layout: key.layout, threshold: key.threshold };
      break;
  }
  return `${prefix}.${encodePayload(JSON.stringify(ordered))}`;
}

function asObject(parsed: unknown, what: string): Record<string, unknown> {
  if (typeof parsed !== 'object' || parsed === null) {
    throw new Error(`slug payload is not a ${what} object`);
  }
  return parsed as Record<string, unknown>;
}

function reqString(obj: Record<string, unknown>, field: string): string {
  const value = obj[field];
  if (typeof value !== 'string') {
    throw new Error(`slug payload field \`${field}\` must be a string`);
  }
  return value;
}

// Mirrors serde's `Option<String>`: an absent field deserializes to `None`
// (here `null`), and a present field must be a string or explicit `null`.
function optString(obj: Record<string, unknown>, field: string): string | null {
  const value = obj[field];
  if (value === undefined || value === null) {
    return null;
  }
  if (typeof value !== 'string') {
    throw new Error(`slug payload field \`${field}\` must be a string or null`);
  }
  return value;
}

// Mirrors serde's `f64` deserialization, which rejects an overflowing JSON
// number literal (for example `1e400`) as "number out of range". JavaScript's
// `JSON.parse` instead overflows to `Infinity`, so without the finiteness check
// a forged threshold of `1e400` would survive decode here while the Rust server
// maps the same payload to the malformed-slug 400 path.
function reqNumber(obj: Record<string, unknown>, field: string): number {
  const value = obj[field];
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new Error(`slug payload field \`${field}\` must be a finite number`);
  }
  return value;
}

// Mirrors serde's `i32` deserialization: the value must be a whole number inside
// the signed-32-bit range. The Rust `ChartKey::QueryMeasurement.query_idx` is an
// `i32`, so `from_slug` rejects a non-integer (`1.5`) or out-of-range (`2**31`)
// payload as a malformed slug. Without this check a forged `query_idx` survives
// decode and only fails later at the Postgres `integer` bind, turning a 400 into
// an unhandled 500.
const I32_MIN = -2_147_483_648;
const I32_MAX = 2_147_483_647;
function reqI32(obj: Record<string, unknown>, field: string): number {
  const value = reqNumber(obj, field);
  if (!Number.isInteger(value) || value < I32_MIN || value > I32_MAX) {
    throw new Error(`slug payload field \`${field}\` must be a 32-bit integer`);
  }
  return value;
}

/**
 * Parse a slug previously produced by [`chartKeyToSlug`]. Validates the full
 * payload shape per variant (not just the `k` discriminant), so a
 * known-discriminant-but-malformed payload is rejected rather than returned as
 * a `ChartKey` with `undefined` required fields. This matches serde's
 * reject-on-bad-shape contract and keeps the cast sound by constructing the
 * result from validated fields.
 */
export function chartKeyFromSlug(slug: string): ChartKey {
  const obj = asObject(decodePayload(slug), 'chart key');
  const k = obj.k;
  if (typeof k !== 'string' || !(CHART_KINDS as readonly string[]).includes(k)) {
    throw new Error(`slug payload has unknown chart key discriminant \`${String(k)}\``);
  }
  const kind = k as ChartKind;
  switch (kind) {
    case 'QueryMeasurement':
      return {
        k: kind,
        dataset: reqString(obj, 'dataset'),
        dataset_variant: optString(obj, 'dataset_variant'),
        scale_factor: optString(obj, 'scale_factor'),
        storage: reqString(obj, 'storage'),
        query_idx: reqI32(obj, 'query_idx'),
      };
    case 'CompressionTime':
    case 'CompressionSize':
      return {
        k: kind,
        dataset: reqString(obj, 'dataset'),
        dataset_variant: optString(obj, 'dataset_variant'),
      };
    case 'RandomAccess':
      return { k: kind, dataset: reqString(obj, 'dataset') };
    case 'VectorSearch':
      return {
        k: kind,
        dataset: reqString(obj, 'dataset'),
        layout: reqString(obj, 'layout'),
        threshold: reqNumber(obj, 'threshold'),
      };
  }
}

/**
 * Build the canonical JSON for a group key with keys in the Rust declaration
 * order (`k` first), then render `<prefix>.<base64url-of-json>`.
 */
export function groupKeyToSlug(key: GroupKey): string {
  const prefix = familyForGroupKind(key.k).groupSlugPrefix;
  let ordered: Record<string, unknown>;
  switch (key.k) {
    case 'QueryGroup':
      ordered = {
        k: key.k,
        dataset: key.dataset,
        dataset_variant: key.dataset_variant,
        scale_factor: key.scale_factor,
        storage: key.storage,
      };
      break;
    case 'VectorSearchGroup':
      ordered = { k: key.k, dataset: key.dataset, layout: key.layout };
      break;
    case 'CompressionTimeGroup':
    case 'CompressionSizeGroup':
    case 'RandomAccessGroup':
      ordered = { k: key.k };
      break;
  }
  return `${prefix}.${encodePayload(JSON.stringify(ordered))}`;
}

/**
 * Parse a slug previously produced by [`groupKeyToSlug`]. Validates the full
 * payload shape per variant (see [`chartKeyFromSlug`] for the rationale).
 */
export function groupKeyFromSlug(slug: string): GroupKey {
  const obj = asObject(decodePayload(slug), 'group key');
  const k = obj.k;
  if (typeof k !== 'string' || !(GROUP_KINDS as readonly string[]).includes(k)) {
    throw new Error(`slug payload has unknown group key discriminant \`${String(k)}\``);
  }
  const kind = k as GroupKind;
  switch (kind) {
    case 'QueryGroup':
      return {
        k: kind,
        dataset: reqString(obj, 'dataset'),
        dataset_variant: optString(obj, 'dataset_variant'),
        scale_factor: optString(obj, 'scale_factor'),
        storage: reqString(obj, 'storage'),
      };
    case 'VectorSearchGroup':
      return { k: kind, dataset: reqString(obj, 'dataset'), layout: reqString(obj, 'layout') };
    case 'CompressionTimeGroup':
    case 'CompressionSizeGroup':
    case 'RandomAccessGroup':
      return { k: kind };
  }
}
